File: promise/src/test/java/com/voxeet/testpromise/all/PromiseTestSimpleAllResolve.java
package com.voxeet.testpromise.all;

import androidx.annotation.NonNull;

import com.voxeet.promise.Promise;
import com.voxeet.promise.solve.PromiseExec;
import com.voxeet.promise.solve.PromiseSolver;
import com.voxeet.promise.solve.Solver;
import com.voxeet.testpromise.utils.AndroidMockUtil;

import org.junit.Before;
import org.junit.Test;

import java.util.Arrays;
import java.util.List;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

public class PromiseTestSimpleAllResolve {

    @Before
    public void setHandler() {
        Promise.setHandler(AndroidMockUtil.mockMainThreadHandler());
    }

    @Test
    public void test() throws InterruptedException {
        final CountDownLatch latch = new CountDownLatch(1);

        final boolean[] catched = {true};
        System.out.println("executing test");

        Promise.all(
                new Promise<>(solver -> {
                    catched[0] = false;
                    solver.resolve("called 1");
                }),
                Promise.resolve("called 2")
        ).then((PromiseExec<List<String>, Void>) (result, solver) -> {
            System.out.println(Arrays.toString(result.toArray()));
            catched[0] = false;
            latch.countDown();
        }).error(error -> {
            error.printStackTrace();
            catched[0] = true;
            latch.countDown();
        });

        //6s are enough
        latch.await(6, TimeUnit.SECONDS);
        if (catched[0]) {
            throw new IllegalStateException("Expected 0 error...");
        }
    }
}

File: promise/src/test/kotlin/com/voxeet/testpromise/all/PromiseTestSimpleAllResolve.kt
package com.voxeet.testpromise.all

import com.voxeet.promise.HandlerFactory
import com.voxeet.promise.Promise
import com.voxeet.promise.await
import com.voxeet.promise.awaitNonNull
import com.voxeet.promise.solve.Solver
import com.voxeet.testpromise.mockedhandler
import com.voxeet.testpromise.registerConfigurationForNoRethrowResolve
import kotlinx.coroutines.test.runTest
import org.junit.Before
import org.junit.Test
import org.junit.jupiter.api.Assertions.assertTrue

class PromiseTestSimpleAllResolve {
    @Before
    fun setHandler() {
        HandlerFactory.setHandler(mockedhandler())
        registerConfigurationForNoRethrowResolve()
    }

    @Test
    fun test() = runTest {
        val called = Promise.all(
            Promise { solver: Solver<String?> ->
                solver.resolve("called 1")
            },
            Promise.resolve("called 2")
        ).then { result: List<String?> ->
            println("execution result = ")
            println(result.toTypedArray().contentToString())
            false
        }.then { !it }.awaitNonNull()

        assertTrue(called)
    }
}
